### ISSUE TITLE
Compiler errors for duplicate modifiers are inconsistent

### ISSUE THREAD

**Issue by DoctorKrolic** [dotnet/roslyn#66077]
Compiler error `CS1004: Duplicate '{0}' modifier` is reported during binding, so if we just parse a syntax tree like below, lets say, via `CSharpSyntaxTree.ParseText`...
```cs
public public class C
{
}
```
... it wouldn't contain any diagnostics at all.

The behaviour is different for statements though. If a stetement contains duplicate modifiers, error `CS1031: Type expected` is reported during parsing on _every_ duplicate modifier. So code like:
```cs
class C
{
    void M()
    {
        const const const int a = 0;
    }
}
```
... produces 2 `CS1031: Type expected` errors on 2 `const` modifiers:
![devenv_owYMsyShh7](https://user-images.githubusercontent.com/70431552/208686664-10fe7988-3c13-4196-8445-33566aedbfd7.png)
_For whatever reason sharplab doesn't properly render diagnostic locations for this case, but if you want to see it threre - [here is the link](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCecAmIA1AD4ACATAAQDCAsAFADejVbVZALFQLIAUASlbsWDduKoBjAPYwAzgimyFS+YtiKIVALxUADAG5hbAL6MTQA)_
No additional errors are reported at the binding stage for statements.

This two mechanisms are mixed together for local functions. Local finctions are statements, so `CS1031: Type expected` is reported during parsing. However, at the binding stage error `CS1004: Duplicate '{0}' modifier` is reported as well, so we have 2 diagnostics on the same token:
```cs
class C
{
    void M()
    {
        static static void M() {}
    }
}
```
[Sharplab link](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCecAmIA1AD4ACATAAQDCAsAFADejVbVZALFQLIAUASlbsWDduI4BGAGxTZXXoKpMAvsLZqGKoA)

Over all, there are several flaws with the current design, described above:
1. No errors about duplicate declaration modifiers after parsing even though parser has all required information to report such errors
2. Not intuitive `CS1031: Type expected` error for statements with duplicate modifiers even when the statement type is specified
3. If there are several duplicate modifiers in a row, for declarations only 1 error is reported on the first duplicate modifier. For statements, there are several errors reported on every duplicate modifier
4. Errors are produced during parsing for statements, but during binding for declarations
5. 2 errors instead of 1 are reported for local finctions. Ideally, it should only be `CS1004: Duplicate '{0}' modifier` (see https://github.com/dotnet/roslyn/issues/32106)

**Comment by CyrusNajmabadi** [comment]
My preference is that all modifier errors get reported in a unified helper in binding.  From a syntactic/grammar perspective, there is nothing wrong, so having in the parser just leads to complexity.
